lib/puppet_x/libkv/file_plugin.rb

```diff
@@ -0,0 +1,315 @@
+      # will be appended to the message by the originating libkv::get()
+      rescue Errno::ENOENT
+        err_msg = "libkv plugin #{@name}: Key not found"
+      rescue Timeout::Error
+        err_msg = "libkv plugin #{@name}: Timed out waiting for key file lock"
+      rescue Exception => e
+        err_msg = "Key retrieval failed: #{e.message}"
+      end
+    end
+    { :result => value, :err_msg => err_msg }
+  end
+
+  # Returns a list of all keys/value pairs in a folder
+  #
+  # This implementation is best effort.  It will attempt to retrieve the
+  # information in a folder and only fail if the folder itself cannot be
+  # accessed.  Individual key retrieval failures will be ignored.
+  #
+  # @return results Hash
+  #   * :result - Hash of retrieved key/value pairs; nil if the
+  #     retrieval operation failed
+  #   * :err_msg - String. Explanatory text upon failure; nil otherwise.
+  #
+  def list(keydir)
+    pairs = nil
+    err_msg = nil
+    dir = File.join(@root_path, keydir)
+    if Dir.exist?(dir)
+      pairs = {}
+      Dir.glob(File.join(dir,'*')).each do |keyfile|
+        key = keyfile.gsub(@root_path + File::SEPARATOR,'')
+        result = get(key)
+        unless result[:result].nil?
+          pairs[key] = result[:result]
+        end
+      end
+    else
+      # Don't need to specify the key folder in the error message, as the key
+      # folder will be reported in the error message generated by the
+      # originating libkv::list()
+       err_msg = "libkv plugin #{@name}: Key folder not found"
+    end
+
+    { :result => pairs, :err_msg => err_msg }
+  end
+
+  # @return unique identifier assigned to this plugin instance
+  def name
+    @name
+  end
+
+  # Sets the data at `key` to a `value` in the configured backend.
+  #
+  # @param key String key
+  # @param value String value
+  #
+  # @return results Hash
+  #   * :result - Boolean indicating whether operation succeeded
+  #   * :err_msg - String. Explanatory text upon failure; nil otherwise.
+  #
+  def put(key, value)
+    success = nil
+    err_msg = nil
+
+    begin
+      # create relative directory for the key file
+      keydir = File.dirname(key)
+      unless keydir == '.'
+        Dir.chdir(@root_path) do
+          FileUtils.mkdir_p(keydir, :mode => 0750)
+          FileUtils.chown_R(@user, @group, keydir) if @user || @group
+        end
+      end
+
+      # create key file
+      key_file = File.join(@root_path, key)
+      Timeout::timeout(@lock_timeout_seconds) do
+        # To ensure all threads are not sharing the same file descriptor
+        # do **NOT** use a File.open block!
+        # Also, don't use 'w' as it truncates file before the lock is obtained
+        file = File.open(key_file, File::RDWR|File::CREAT, 0640)
+        file.flock(File::LOCK_EX)
+        file.rewind
+        file.write(value)
+        file.flush
+        file.truncate(file.pos)
+        file.close # lock released with close
```

Comment:
We'll probably need to throw this into an `ensure` of some sort in case we hit the `Timeout` exception.